### ISSUE TITLE
Kinetic Merge faults with java.util.NoSuchElementException: empty.head

### ISSUE THREAD

**Issue by sageserpent-open** [sageserpent-open/kineticMerge#124]
Seen on commit 42d2deb46c5cef6945b579ccaff00c086e5cbaa8.

Merging the Kinetic Merge repository, from commit 850d1885301d26d439167563efb044ba0d91e291 into commit 42d2deb46c5cef6945b579ccaff00c086e5cbaa8 (using a build from that commit too).

Full stacktrace:

```text
java.util.NoSuchElementException: empty.head
	at scala.collection.immutable.Vector.head(Vector.scala:288)
	at com.sageserpent.kineticmerge.core.MoveDestinationsReport$.$anonfun$6$$anonfun$6(MoveDestinationsReport.scala:303)
	at scala.collection.StrictOptimizedIterableOps.flatMap(StrictOptimizedIterableOps.scala:118)
	at scala.collection.StrictOptimizedIterableOps.flatMap$(StrictOptimizedIterableOps.scala:105)
	at scala.collection.immutable.Set$Set1.flatMap(Set.scala:165)
	at com.sageserpent.kineticmerge.core.MoveDestinationsReport$.$anonfun$6(MoveDestinationsReport.scala:296)
	at scala.collection.immutable.List.flatMap(List.scala:294)
	at scala.collection.immutable.List.flatMap(List.scala:79)
	at com.sageserpent.kineticmerge.core.MoveDestinationsReport$.evaluateSpeculativeSourcesAndDestinations(MoveDestinationsReport.scala:209)
	at com.sageserpent.kineticmerge.core.CodeMotionAnalysisExtension$.merge(CodeMotionAnalysisExtension.scala:257)
	at com.sageserpent.kineticmerge.Main$InWorkingDirectory.indexUpdates$$anonfun$2$$anonfun$2(Main.scala:1265)
	at cats.data.EitherT.bimap$$anonfun$1(EitherT.scala:423)
	at cats.data.WriterT.map$$anonfun$1(WriterT.scala:138)
	at apply$extension @ com.sageserpent.kineticmerge.Main$.com$sageserpent$kineticmerge$Main$$$right(Main.scala:392)
	at apply @ com.sageserpent.kineticmerge.Main$.com$sageserpent$kineticmerge$Main$$$labelExceptionWith(Main.scala:398)
	at apply @ com.sageserpent.kineticmerge.Main$.com$sageserpent$kineticmerge$Main$$$labelExceptionWith(Main.scala:398)
	at flatMap @ com.sageserpent.kineticmerge.Main$.mergeTheirBranch(Main.scala:371)
	at apply$extension @ com.sageserpent.kineticmerge.Main$.com$sageserpent$kineticmerge$Main$$$right(Main.scala:392)
	at apply @ com.sageserpent.kineticmerge.Main$.com$sageserpent$kineticmerge$Main$$$labelExceptionWith(Main.scala:398)
	at foldM @ com.sageserpent.kineticmerge.Main$InWorkingDirectory.indexUpdates(Main.scala:1243)
	at foldM @ com.sageserpent.kineticmerge.Main$InWorkingDirectory.indexUpdates(Main.scala:1243)
	at apply$extension @ com.sageserpent.kineticmerge.Main$.com$sageserpent$kineticmerge$Main$$$right(Main.scala:392)
	at apply @ com.sageserpent.kineticmerge.Main$.com$sageserpent$kineticmerge$Main$$$labelExceptionWith(Main.scala:398)
	at apply @ com.sageserpent.kineticmerge.Main$.com$sageserpent$kineticmerge$Main$$$labelExceptionWith(Main.scala:398)
	at foldM @ com.sageserpent.kineticmerge.Main$InWorkingDirectory.indexUpdates(Main.scala:1243)
	at foldM @ com.sageserpent.kineticmerge.Main$InWorkingDirectory.indexUpdates(Main.scala:1243)
	at apply$extension @ com.sageserpent.kineticmerge.Main$.com$sageserpent$kineticmerge$Main$$$right(Main.scala:392)
	at apply @ com.sageserpent.kineticmerge.Main$.com$sageserpent$kineticmerge$Main$$$labelExceptionWith(Main.scala:398)

```

**Comment by sageserpent-open** [comment]
Also reproduced on commit 34a0ff8b1d77a42146f193d264b1df80dd8b6238, prior to work in progress on #35 (where this was orginally seen).

**Comment by sageserpent-open** [comment]
My hypothesis is that some of the fundamental migration scenarios aren't supported - in this case, the code in `MoveDestinationsReport.evaluateSpeculativeSourcesAndDestinations` is misunderstanding a [migrated deletion where the source of the move is replaced by some unrelated content on the same side as the move.](https://github.com/sageserpent-open/kineticMerge/blob/main/documents/designNotes/migratingADeletionWithAnEditOnTheMoveSide.excalidraw.svg).

It was useful writing lots of elementary and systematically-enumerated test cases for migration through file renames in `CodeMotionAnalysisExtensionTest`, both as a pedagogical aid in terms of specification-by-example and as a means to ensure coverage of scenarios.

Time to cover the various migration scenarios outlined in the Exclidraw documents.

**Comment by sageserpent-open** [comment]
Evidence of successful merge via build from commit 1572d75ed4180f7dd92e7e93e00a1cbf613c6c0f:

```text
kinetic-merge --no-commit 850d188                                                                                 
SLF4J(I): Connected with provider of type [ch.qos.logback.classic.spi.LogbackServiceProvider]
                                                                                                                                                                                                                                                           
Single move of:
Section(
  label = "BASE: 34a0ff8b",
  path = /Users/gerardmurphy/IdeaProjects/kineticMerge-dogfood-examples/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala,
  start = TextPosition(line = 1183, characterOffset = 18),
  onePastEnd = TextPosition(line = 1187, characterOffset = 20),
  startTokenIndex = 6024,
  sizeInTokens = 12,
  content summary = """matchingFingerprintsAcrossSides(
                    baseFingerprints,
                    leftFingerprints ... ,
                    rightFingerprints.tail,
                    """
)
to:
Section(
  label = "THEIRS: 850d188",
  path = /Users/gerardmurphy/IdeaProjects/kineticMerge-dogfood-examples/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala,
  start = TextPosition(line = 1249, characterOffset = 18),
  onePastEnd = TextPosition(line = 1253, characterOffset = 20),
  startTokenIndex = 6270,
  sizeInTokens = 12,
  content summary = """matchingFingerprintsAcrossSides(
                    baseFingerprints,
                    leftFingerprints ... ,
                    rightFingerprints.tail,
                    """
).
Single move of:
Section(
  label = "BASE: 34a0ff8b",
  path = /Users/gerardmurphy/IdeaProjects/kineticMerge-dogfood-examples/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala,
  start = TextPosition(line = 1005, characterOffset = 18),
  onePastEnd = TextPosition(line = 1009, characterOffset = 20),
  startTokenIndex = 5259,
  sizeInTokens = 14,
  content summary = """matchingFingerprintsAcrossSides(
                    baseFingerprints.tail ... ,
                    rightFingerprints.tail,
                    """
)
to:
Section(
  label = "THEIRS: 850d188",
  path = /Users/gerardmurphy/IdeaProjects/kineticMerge-dogfood-examples/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala,
  start = TextPosition(line = 1048, characterOffset = 18),
  onePastEnd = TextPosition(line = 1052, characterOffset = 20),
  startTokenIndex = 5417,
  sizeInTokens = 14,
  content summary = """matchingFingerprintsAcrossSides(
                    baseFingerprints.tail ... ,
                    rightFingerprints.tail,
                    """
).
Single move of:
Section(
  label = "BASE: 34a0ff8b",
  path = /Users/gerardmurphy/IdeaProjects/kineticMerge-dogfood-examples/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala,
  start = TextPosition(line = 1069, characterOffset = 20),
  onePastEnd = TextPosition(line = 1069, characterOffset = 23),
  startTokenIndex = 5530,
  sizeInTokens = 1,
  content = "if "
)
to:
Section(
  label = "OURS: 42d2deb46c5cef6945b579ccaff00c086e5cbaa8",
  path = /Users/gerardmurphy/IdeaProjects/kineticMerge-dogfood-examples/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala,
  start = TextPosition(line = 1055, characterOffset = 16),
  onePastEnd = TextPosition(line = 1055, characterOffset = 19),
  startTokenIndex = 5462,
  sizeInTokens = 1,
  content = "if "
).
Single move of:
Section(
  label = "BASE: 34a0ff8b",
  path = /Users/gerardmurphy/IdeaProjects/kineticMerge-dogfood-examples/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala,
  start = TextPosition(line = 1124, characterOffset = 18),
  onePastEnd = TextPosition(line = 1128, characterOffset = 20),
  startTokenIndex = 5771,
  sizeInTokens = 12,
  content summary = """matchingFingerprintsAcrossSides(
                    baseFingerprints.tail ... ,
                    rightFingerprints.tail,
                    """
)
to:
Section(
  label = "THEIRS: 850d188",
  path = /Users/gerardmurphy/IdeaProjects/kineticMerge-dogfood-examples/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala,
  start = TextPosition(line = 1182, characterOffset = 18),
  onePastEnd = TextPosition(line = 1186, characterOffset = 20),
  startTokenIndex = 5987,
  sizeInTokens = 12,
  content summary = """matchingFingerprintsAcrossSides(
                    baseFingerprints.tail ... ,
                    rightFingerprints.tail,
                    """
).
Single move of:
Section(
  label = "BASE: 34a0ff8b",
  path = /Users/gerardmurphy/IdeaProjects/kineticMerge-dogfood-examples/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala,
  start = TextPosition(line = 1012, characterOffset = 24),
  onePastEnd = TextPosition(line = 1013, characterOffset = 20),
  startTokenIndex = 5280,
  sizeInTokens = 9,
  content = """.map(Match.AllSides.apply)
                    """
)
to:
Section(
  label = "OURS: 42d2deb46c5cef6945b579ccaff00c086e5cbaa8",
  path = /Users/gerardmurphy/IdeaProjects/kineticMerge-dogfood-examples/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala,
  start = TextPosition(line = 1002, characterOffset = 38),
  onePastEnd = TextPosition(line = 1003, characterOffset = 16),
  startTokenIndex = 5246,
  sizeInTokens = 9,
  content = """.map(Match.AllSides.apply)
                """
).
Single move of:
Section(
  label = "BASE: 34a0ff8b",
  path = /Users/gerardmurphy/IdeaProjects/kineticMerge-dogfood-examples/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala,
  start = TextPosition(line = 1187, characterOffset = 20),
  onePastEnd = TextPosition(line = 1187, characterOffset = 23),
  startTokenIndex = 6036,
  sizeInTokens = 1,
  content = "if "
)
to:
Section(
  label = "OURS: 42d2deb46c5cef6945b579ccaff00c086e5cbaa8",
  path = /Users/gerardmurphy/IdeaProjects/kineticMerge-dogfood-examples/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala,
  start = TextPosition(line = 1163, characterOffset = 16),
  onePastEnd = TextPosition(line = 1163, characterOffset = 19),
  startTokenIndex = 5912,
  sizeInTokens = 1,
  content = "if "
).
Single move of:
Section(
  label = "BASE: 34a0ff8b",
  path = /Users/gerardmurphy/IdeaProjects/kineticMerge-dogfood-examples/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala,
  start = TextPosition(line = 1065, characterOffset = 18),
  onePastEnd = TextPosition(line = 1069, characterOffset = 20),
  startTokenIndex = 5518,
  sizeInTokens = 12,
  content summary = """matchingFingerprintsAcrossSides(
                    baseFingerprints.tail ... .tail,
                    rightFingerprints,
                    """
)
to:
Section(
  label = "THEIRS: 850d188",
  path = /Users/gerardmurphy/IdeaProjects/kineticMerge-dogfood-examples/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala,
  start = TextPosition(line = 1115, characterOffset = 18),
  onePastEnd = TextPosition(line = 1119, characterOffset = 20),
  startTokenIndex = 5704,
  sizeInTokens = 12,
  content summary = """matchingFingerprintsAcrossSides(
                    baseFingerprints.tail ... .tail,
                    rightFingerprints,
                    """
).
Single move of:
Section(
  label = "BASE: 34a0ff8b",
  path = /Users/gerardmurphy/IdeaProjects/kineticMerge-dogfood-examples/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala,
  start = TextPosition(line = 1128, characterOffset = 20),
  onePastEnd = TextPosition(line = 1128, characterOffset = 23),
  startTokenIndex = 5783,
  sizeInTokens = 1,
  content = "if "
)
to:
Section(
  label = "OURS: 42d2deb46c5cef6945b579ccaff00c086e5cbaa8",
  path = /Users/gerardmurphy/IdeaProjects/kineticMerge-dogfood-examples/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala,
  start = TextPosition(line = 1109, characterOffset = 16),
  onePastEnd = TextPosition(line = 1109, characterOffset = 19),
  startTokenIndex = 5687,
  sizeInTokens = 1,
  content = "if "
).
Successful merge, leaving merged changes in the index for review...

```

**Comment by sageserpent-open** [comment]
Merged on to `main` (using Kinetic Merge!) in ef56f6df3aab47d46867217f7146fc1b74ffe839.

**Comment by sageserpent-open** [comment]
This went out in Release 1.4.0.
